src/freertos_drivers/esp32/Esp32WiFiManager.cxx

```diff
@@ -0,0 +1,985 @@
+            // active connections since they will be invalid until a new IP
+            // has been provisioned.
+            wifi->stop_hub();
+            wifi->stop_uplink();
+
+            // Make sure we don't try and reload configuration since we can't
+            // create outbound connections at this time.
+            wifi->configReloadRequested_ = false;
+        }
+
+        // Check if there are configuration changes to pick up.
+        if(wifi->configReloadRequested_)
+        {
+            // Since we are loading configuration data, shutdown the hub and
+            // uplink if created previously.
+            wifi->stop_hub();
+            wifi->stop_uplink();
+
+            if(CDI_READ_TRIMMED(wifi->cfg_.sleep, wifi->configFd_))
+            {
+                // When sleep is enabled this will trigger the WiFi system to
+                // only wake up every DTIM period to receive beacon updates.
+                // no data loss is expected for this setting but it does delay
+                // receiption until the DTIM period.
+                ESP_ERROR_CHECK(esp_wifi_set_ps(WIFI_PS_MIN_MODEM));
+            }
+            else
+            {
+                // When sleep is disabled the WiFi radio will always be active.
+                // This will increase power consumption of the ESP32 but it
+                // will result in a more reliable behavior when the ESP32 is
+                // connected to an always-on power supply (ie: not a battery).
+                ESP_ERROR_CHECK(esp_wifi_set_ps(WIFI_PS_NONE));
+            }
+
+            if(CDI_READ_TRIMMED(wifi->cfg_.hub().enable, wifi->configFd_))
+            {
+                // Since hub mode is enabled start the HUB creation process.
+                wifi->start_hub();
+            }
+            // Start the uplink connection process in the background.
+            wifi->start_uplink();
+            wifi->configReloadRequested_ = false;
+        }
+
+        // Sleep until we are woken up again for configuration update or WiFi
+        // event.
+        ulTaskNotifyTake(pdTRUE, portMAX_DELAY);
+    }
+
+    return nullptr;
+}
+
+// Shuts down the hub listener (if enabled and running) for this node.
+void Esp32WiFiManager::stop_hub()
+{
+    if(hub_)
+    {
+        mdns_unpublish(hubServiceName_.c_str());
+        LOG(INFO, "[HUB] Shutting down TCP/IP listener");
+        //hub_->shutdown();
+        hub_.reset(nullptr);
+    }
+}
+
+// Creates a hub listener for this node after loading configuration details.
+void Esp32WiFiManager::start_hub()
+{
+    hubServiceName_ = cfg_.hub().service_name().read(configFd_);
+    uint16_t hub_port = CDI_READ_TRIMMED(cfg_.hub().port, configFd_);
+
+    LOG(INFO, "[HUB] Starting TCP/IP listener on port %d", hub_port);
+    hub_.reset(new GcTcpHub(stack_->can_hub(), hub_port));
+
+    // wait for the hub to complete it's startup tasks
+    while(!hub_->is_started())
+    {
+        vTaskDelay(HUB_STARTUP_DELAY);
```

Comment:
this should be usleep instead (eventually to migrate to nanosleep)

Reply:
will switch and update HUB_STARTUP_DELAY to use MSEC_TO_USEC instead.